runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/GrpcDirectCommitWorkStream.java

```diff
@@ -0,0 +1,243 @@
+  private final WeightedBoundedQueue<Commit> commitQueue;
+  private final ExecutorService commitSender;
+  private final Consumer<Commit> onFailedCommits;
+  private final AtomicLong activeCommitBytes;
+
+  private GrpcDirectCommitWorkStream(
+      Function<StreamObserver<StreamingCommitResponse>, StreamObserver<StreamingCommitWorkRequest>>
+          startCommitWorkRpcFn,
+      BackOff backoff,
+      StreamObserverFactory streamObserverFactory,
+      Set<AbstractWindmillStream<?, ?>> streamRegistry,
+      int logEveryNStreamFailures,
+      ThrottleTimer commitWorkThrottleTimer,
+      JobHeader jobHeader,
+      AtomicLong idGenerator,
+      int streamingRpcBatchLimit,
+      Consumer<Commit> onFailedCommits) {
+    super(
+        startCommitWorkRpcFn,
+        backoff,
+        streamObserverFactory,
+        streamRegistry,
+        logEveryNStreamFailures,
+        commitWorkThrottleTimer,
+        jobHeader,
+        idGenerator,
+        streamingRpcBatchLimit);
+    this.commitQueue =
+        WeightedBoundedQueue.create(
+            MAX_COMMIT_QUEUE_BYTES, commit -> Math.min(MAX_COMMIT_QUEUE_BYTES, commit.getSize()));
+    this.commitSender =
+        Executors.newSingleThreadScheduledExecutor(
+            new ThreadFactoryBuilder()
+                .setNameFormat("DirectCommitWorkSenderThread")
+                .setUncaughtExceptionHandler(
+                    (t, e) ->
+                        LOG.error(
+                            "{} failed due to uncaught exception during execution. ",
+                            t.getName(),
+                            e))
+                .build());
+    this.onFailedCommits = onFailedCommits;
+    this.activeCommitBytes = new AtomicLong();
+  }
+
+  public static GrpcDirectCommitWorkStream create(
+      Function<StreamObserver<StreamingCommitResponse>, StreamObserver<StreamingCommitWorkRequest>>
+          startCommitWorkRpcFn,
+      BackOff backoff,
+      StreamObserverFactory streamObserverFactory,
+      Set<AbstractWindmillStream<?, ?>> streamRegistry,
+      int logEveryNStreamFailures,
+      ThrottleTimer commitWorkThrottleTimer,
+      JobHeader jobHeader,
+      AtomicLong idGenerator,
+      int streamingRpcBatchLimit,
+      Consumer<Commit> onFailedCommits) {
+    GrpcDirectCommitWorkStream commitWorkStream =
+        new GrpcDirectCommitWorkStream(
+            startCommitWorkRpcFn,
+            backoff,
+            streamObserverFactory,
+            streamRegistry,
+            logEveryNStreamFailures,
+            commitWorkThrottleTimer,
+            jobHeader,
+            idGenerator,
+            streamingRpcBatchLimit,
+            onFailedCommits);
+    commitWorkStream.startStream();
+    commitWorkStream.startCommitSender();
+    return commitWorkStream;
+  }
+
+  @Override
+  public void queueCommit(
+      WorkItemCommitRequest workItemCommitRequest, ComputationState computationState, Work work) {
```

Comment:
instead of ComputationState how about Consumer bound to computationState by caller?
Instead of just failedCommit could have commitCompleteConsumer taking a commit and Windmill.CommitStatus.

Seems like it would be easier to test in isolation.

Reply:
Added another consumer for onCommitComplete vs reusing failed commits

I was a bit unsure of what it would look like to use a Consumer bound to computation state by the caller, can you clarify? Thanks!

Reply:
I meant that this class could avoid having to know about ComputationState, that the 
`state.completeWorkAndScheduleNextWorkForKey(`
calls could live within the failedConsumer or commitCompleteConsumer.

ie something like this:
```
Consumer<CompletedCommit> consumer = (CompletedCommit c) -> {  state.completeWorkAndScheduleNextWorkForKey(....); }
```

Looks like you might have to pass in the computationId with the commit.